text/0000-native-lifecycle-mixed-mode.md

```diff
@@ -0,0 +1,152 @@
+
+To resolve this, this RFC proposes a new "native lifecycle mixed mode," inspired by [mixed shadow DOM mode](https://github.com/salesforce/lwc-rfcs/blob/master/text/0115-mixed-shadow-mode.md). It would work in much the same way, allowing components to gradually opt-in to native custom element lifecycle, while also maintaining consistency within the DOM tree for that component.
+
+## Basic example
+
+Opting in:
+
+```js
+export default class extends LightningElement {
+  static lifecycleSupportMode = 'native';
+}
+```
+
+Similar to shadow DOM mixed mode, you can also opt out:
+
+```js
+export default class extends LightningElement {
+  static lifecycleSupportMode = 'reset';
+}
+```
+
+Like shadow DOM mixed mode, this `'reset'` is mostly useful if you have a superclass that declares `'native'`, which you would like to override.
+
+## Motivation
+
+The synthetic lifecycle polyfill is technical debt. If LWC were designed from scratch today, it wouldn't ship with its own implementation of `connectedCallback` or `disconnectedCallback` – it would just defer to the native browser behavior.
+
+Unfortunately, due to subtle differences between native and synthetic lifecycle, and due to the Lightning Platform's backwards compatibility guarantees, it is not trivial to swap one implementation out for the other.
+
+At the same time, component authors often complain about bugs in synthetic lifecycle ([and](https://github.com/salesforce/lwc/issues/2609) [there](https://github.com/salesforce/lwc/issues/3361) [are](https://github.com/salesforce/lwc/issues/1452) [many](https://github.com/salesforce/lwc/issues/3823)). The most common complaint is that `disconnectedCallback` does not consistently fire, which can lead to memory leaks or buggy behavior. Unfortunately, fixing these bugs would necessarily create observable behavior that would break some other component that relies on the bug (see [Hyrum's Law](https://www.hyrumslaw.com/)).
+
+To unblock component authors who _do_ want non-buggy behavior for `connectedCallback`/`disconnectedCallback`, we should allow them to opt-in to native lifecycle. This paves the way for a polyfill-free future, while still preserving backwards compatibility and giving a migration path for authors who need time to upgrade.
+
+## Detailed design
+
+### The `lifecycleSupportMode` property
+
+Like `shadowSupportMode`, `lifecycleSupportMode` has two possible values:
+
+- `'native'`
+- `'reset'` (default, i.e. synthetic mode)
+
+Any other value throws an error at runtime.
+
+### Polyfill behavior
+
+Similar to [`@lwc/synthetic-shadow`](https://www.npmjs.com/package/@lwc/synthetic-shadow) and [`@lwc/aria-reflection`](https://www.npmjs.com/package/@lwc/aria-reflection), we will isolate the code necessary for synthetic lifecycle into a standalone package: `@lwc/synthetic-lifecycle`. The polyfill can be loaded using simply:

```

Comment:
Did we consider moving synthetic lifecycle into `@lwc/synthetic-shadow`? Seems like a reasonable fit that wouldn't cause disruption.

Reply:
I considered conflating the two, but I thought it might be best to keep them separate so we can test them separately. Same deal with `@lwc/aria-reflection` (yet another polyfill, loaded by default in LEX).

Reply:
Does it make sense to test them separately? Conflating the two would preclude an app from using `@lwc/synthetic-shadow` with default native lifecycle, but I wonder what the demand for such a configuration is.

Reply:
I guess I'm thinking that component authors may have more trouble with one than the other. In particular, it feels intuitively like it should be "easier" to move to native lifecycle than to native shadow. That said, we ideally want to move the needle on both.

Reply:
I thought about this – I'm not sure we can conflate the two together, because that would be potentially a breaking change for existing components that have `shadowSupportMode = 'native'`. Their timing would certainly change in native lifecycle.